src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -226,19 +236,88 @@ void Instance::InvokeCommand(HandlerContext & handlerContext)
                 [this](HandlerContext & ctx, const auto & commandData) { HandleRequestConstraintBasedForecast(ctx, commandData); });
         }
         return;
+    case CancelRequest::Id:
+        if (!HasFeature(Feature::kStartTimeAdjustment) && !HasFeature(Feature::kForecastAdjustment) &&
+            !HasFeature(Feature::kConstraintBasedAdjustment))
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath, Status::UnsupportedCommand);
+        }
+        else
+        {
+            HandleCommand<CancelRequest::DecodableType>(
+                handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleCancelRequest(ctx, commandData); });
+        }
+        return;
     }
 }
 
-void Instance::HandlePowerAdjustRequest(HandlerContext & ctx, const Commands::PowerAdjustRequest::DecodableType & commandData)
+Status Instance::CheckOptOutAllowsRequest(AdjustmentCauseEnum adjustmentCause)
 {
-    int64_t power        = commandData.power;
-    uint32_t durationSec = commandData.duration;
-    bool validArgs       = false;
-    Status status        = Status::Success;
+    OptOutStateEnum optOutState = mDelegate.GetOptOutState();
+
+    if (adjustmentCause >= AdjustmentCauseEnum::kUnknownEnumValue)
+    {
+        ChipLogError(Zcl, "DEM: adjustment cause is invalid (%d)", static_cast<int>(adjustmentCause));
+        return Status::InvalidValue;
+    }
+
+    switch (optOutState)
+    {
+    case OptOutStateEnum::kNoOptOut: /* User has NOT opted out so allow it */
+        ChipLogProgress(Zcl, "DEM: OptOutState = kNoOptOut");
+        return Status::Success;
+
+    case OptOutStateEnum::kLocalOptOut: /* User has opted out from Local only*/
+        ChipLogProgress(Zcl, "DEM: OptOutState = kLocalOptOut");
+        switch (adjustmentCause)
+        {
+        case AdjustmentCauseEnum::kGridOptimization:
+            return Status::Success;
+        case AdjustmentCauseEnum::kLocalOptimization:
+        default:
```

Comment:
Not having a default (but having a case for kUnknownEnumValue, which you already checked above, I know) will make this not compile when a new cause value is added, which might be desirable.  Similar in other switch statements.

Reply:
Ok will amend